### PR TITLE
taskid and task: {id} validation

### PR DESCRIPTION
**Pull request by tallieber**
…. Add to playbook and test playbook validations

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Validate that taksid field and task: {id} are equals and from uuid type

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: demisto_sdk/commands/common/errors.py [hunk 1]
```diff
@@ -1203,6 +1205,18 @@ def integration_not_runnable():
...
+    @error_code_decorator
+    def invalid_uuid(task_key, id, taskid):
+        return f"On task with the following key: {task_key}, the field of taskid: {taskid} and the feild of id: " \
+               f"{id} must be from uuid type."
+
+    @staticmethod
+    @error_code_decorator
+    def taskid_different_from_id(task_key, id, taskid):
+        return f"On task with the following key: {task_key}, the field of taskid: {taskid} and the feild of id: " \
+               f"{id} must be with equal value. "
```
Comment (bakatzir) [inline demisto_sdk/commands/common/errors.py:1218]: please use the `self.file_path` in the err msg so the user will know in which PB the error is.

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 2]
```diff
@@ -52,6 +54,8 @@ def is_valid_playbook(self, is_new_playbook: bool = True, validate_rn: bool = Tr
                 self.is_delete_context_all_in_playbook(),
                 self.are_tests_configured(),
                 self.is_script_id_valid(id_set_file),
+                self._is_id_uuid(),
+                self._is_taskid_equals_id()
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/playbook.py:58]: follow PEP https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas
```suggestion
                self._is_taskid_equals_id(),
```

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 3]
```diff
@@ -38,6 +38,8 @@ def is_valid_playbook(self, is_new_playbook: bool = True, validate_rn: bool = Tr
                 self.are_tests_configured(),
                 self.is_valid_as_deprecated(),
                 self.is_script_id_valid(id_set_file),
+                self._is_id_uuid(),
+                self._is_taskid_equals_id()
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/playbook.py:42]: follow PEP https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas
add comma

Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/playbook.py:42]: follow PEP https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas
add comma

Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/playbook.py:42]: follow PEP https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas
```suggestion
                self._is_taskid_equals_id()`
```

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 4]
```diff
@@ -356,3 +360,40 @@ def check_script_name(self, pb_script_name, id_set_scripts):
         return any(
             [pb_script_name == id_set_dict[key].get('name') for id_set_dict in id_set_scripts
              for key in id_set_dict])
+
+    def _is_id_uuid(self):
+        """
+        Check that taskid field and id field under task field are both from uuid type
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/playbook.py:366]:
```suggestion
        Check that the taskid field and the  id field under the task field are both in uuid format
```

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 5]
```diff
@@ -356,3 +360,40 @@ def check_script_name(self, pb_script_name, id_set_scripts):
...
+        tasks: dict = self.current_file.get('tasks', {})
+        for task_key, task in tasks.items():
+            taskid = task.get('taskid', '')
+            inner_id = task.get('task', {}).get('id', '')
+            is_valid = is_string_uuid(taskid) and is_string_uuid(inner_id)
+
+            if not is_valid:
+                error_message, error_code = Errors.invalid_uuid(task_key, taskid, inner_id)
+                if self.handle_error(error_message, error_code, file_path=self.file_path):
+                    return is_valid
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/playbook.py:379]: it is clearer to break the loop and have a single exist(return) point in the func
```suggestion
            if not is_valid:
                error_message, error_code = Errors.invalid_uuid(task_key, taskid, inner_id)
                if self.handle_error(error_message, error_code, file_path=self.file_path):
                    break
```

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 6]
```diff
@@ -356,3 +360,40 @@ def check_script_name(self, pb_script_name, id_set_scripts):
...
+        tasks: dict = self.current_file.get('tasks', {})
+        for task_key, task in tasks.items():
+            taskid = task.get('taskid', '')
+            inner_id = task.get('task', {}).get('id', '')
+            is_valid = (taskid == inner_id)
+
+            if not is_valid:
+                error_message, error_code = Errors.taskid_different_from_id(task_key, taskid, inner_id)
+                if self.handle_error(error_message, error_code, file_path=self.file_path):
+                    return is_valid
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/playbook.py:398]: same

File: demisto_sdk/commands/common/hook_validations/test_playbook.py [hunk 7]
```diff
@@ -1,12 +1,22 @@
...
+from demisto_sdk.commands.common.errors import Errors
+from demisto_sdk.commands.common.tools import is_string_uuid
 
 
 class TestPlaybookValidator(ContentEntityValidator):
     """TestPlaybookValidator is designed to validate the correctness of the file structure we enter to content repo for
     both test playbooks and scripts.
     """
 
+    def is_valid_test_playbook(self, validate_rn: bool = True):
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/test_playbook.py:12]: add docstring and return type

File: demisto_sdk/commands/common/hook_validations/test_playbook.py [hunk 8]
```diff
@@ -22,3 +32,40 @@ def is_valid_version(self):  # type: () -> bool
...
+        tasks: dict = self.current_file.get('tasks', {})
+        for task_key, task in tasks.items():
+            taskid = task.get('taskid', '')
+            inner_id = task.get('task', {}).get('id', '')
+            is_valid = is_string_uuid(taskid) and is_string_uuid(inner_id)
+
+            if not is_valid:
+                error_message, error_code = Errors.invalid_uuid(task_key, taskid, inner_id)
+                if self.handle_error(error_message, error_code, file_path=self.file_path):
+                    return is_valid
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/test_playbook.py:51]: same

File: demisto_sdk/commands/common/hook_validations/test_playbook.py [hunk 9]
```diff
@@ -22,3 +32,40 @@ def is_valid_version(self):  # type: () -> bool
...
+        tasks: dict = self.current_file.get('tasks', {})
+        for task_key, task in tasks.items():
+            taskid = task.get('taskid', '')
+            inner_id = task.get('task', {}).get('id', '')
+            is_valid = (taskid == inner_id)
+
+            if not is_valid:
+                error_message, error_code = Errors.taskid_different_from_id(task_key, taskid, inner_id)
+                if self.handle_error(error_message, error_code, file_path=self.file_path):
+                    return is_valid
```
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/test_playbook.py:70]: same

File: demisto_sdk/tests/test_tools.py [hunk 10]
```diff
@@ -49,3 +49,18 @@ def test_get_content_file_type_dump(file_path: str, expected: Callable[[str], st
...
+    Given
+    - A string to check
+
+    When
+    - Running the method 'is_id_uuid' on 'is_valid_playbook'
+
+    Then
+    -  Ensure the method return True or False if the string match uuid or not
+
+    """
```
Comment (bakatzir) [inline demisto_sdk/tests/test_tools.py:66]: I guess the missing test is a commit issue?

File: demisto_sdk/commands/common/hook_validations/test_playbook.py (hunk 7 above)
Comment (bakatzir) [inline demisto_sdk/commands/common/hook_validations/test_playbook.py:12]: TPB default val of rn should be false, as customers do not care when they are changed.
```suggestion
    def is_valid_test_playbook(self, validate_rn: bool = False):
```

**Review by bakatzir — CHANGES_REQUESTED** [review]
Nice. please review my comment

File: demisto_sdk/commands/common/hook_validations/test_playbook.py (hunk 7 above)
Comment (tallieber) [reply demisto_sdk/commands/common/hook_validations/test_playbook.py:12]: OK you are right. also when this function is called the given value of this argument is false

File: demisto_sdk/tests/test_tools.py (hunk 10 above)
Comment (bakatzir) [reply demisto_sdk/tests/test_tools.py:66]: decided by both to rm this test.

File: demisto_sdk/commands/common/hook_validations/test_playbook.py (hunk 7 above)
Comment (bakatzir) [reply demisto_sdk/commands/common/hook_validations/test_playbook.py:12]: Indeed. but to not make this func prone to errors, best to give default vals of the most common usage, which in this case is `False`.

**Review by bakatzir — APPROVED** [review]

**Review by bakatzir — COMMENTED** [review]
please add a line in changelog.md

**Pull request merged**